lib/src/command/login.dart

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import '../command.dart';
+import '../http.dart';
+import '../log.dart' as log;
+import '../oauth2.dart' as oauth2;
+
+/// Handles the `login` pub command.
+class LoginCommand extends PubCommand {
+  @override
+  String get name => 'login';
+  @override
+  String get description => 'Log into pub.dev.';
+  @override
+  String get invocation => 'pub login';
+
+  LoginCommand();
+
+  @override
+  Future run() async {
+    final credentials = oauth2.loadCredentials(cache);
+    if (credentials == null) {
+      final userInfo = retrieveUserInfo();
+      log.message('You are now logged in as $userInfo');
+    } else {
+      final userInfo = await retrieveUserInfo();
+      if (userInfo == null) {
+        log.warning('Your credentials seems broken.\n'
+            'Run `pub logout` to delete your credentials  and try again.');
+      }
+      log.warning('You are already logged in as $userInfo\n'
+          'Run `pub logout` to log out and try again.');
+    }
+  }
+
+  Future<_UserInfo> retrieveUserInfo() async {
+    return await oauth2.withClient(cache, (client) async {
+      final discovery = await httpClient
+          .get('https://accounts.google.com/.well-known/openid-configuration');
```

Comment:
I know that `https://www.googleapis.com/oauth2/v1/userinfo` can return a user profile too.. but I can't find documentation for the API end-point :see_no_evil: